### PR TITLE
Serve command fixes

### PR DESCRIPTION
**Pull request by kspearrin**
## Type of change

- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Resolve open Asana issues related to new serve command.

## Code changes

Issues related to commands handing were mostly related to asynchronous recursive loops that were fixed with https://github.com/bitwarden/jslib/commit/fc0638a7d95e9e678cc125d5186c1dcd8fdad7e6 .

This change also transitions from express.js to koa (made by the same developers as express) for the web server for better async/await support.

The decision was made to do away with the `Session` header option since serve commands are not completely stateless (things are cached within the process memory, for example).

Finally, I added some checks to make sure errors around certain conditions are handled more gracefully. Namely, checking to make sure the vault is in an unlocked state before trying to execute dependent commands, such as `list`.

## Testing requirements

Test all commands and options related to the HTTP server exposed by the `serve` command.

## Before you submit

- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [x] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/commands/serve.command.ts [hunk 1]
```diff
@@ -144,158 +149,263 @@ export class ServeCommand {
...
 
-    server.put("/:object/:id", async (req, res) => {
+    router.put("/:object/:id", async (ctx, next) => {
+      if (await this.errorIfLocked(ctx.response)) {
+        await next();
+        return;
+      }
+      if (this.alreadyRespondedTo(ctx.response)) {
+        return;
+      }
```
Comment (MGibson1) [inline src/commands/serve.command.ts:312]: I'm confused about this helper. Why would response already have a body? Is it from earlier in the middleware cascade?

If so, why don't all of the commands preface with this?

File: package.json [hunk 2]
```diff
@@ -93,18 +96,22 @@
...
     "https-proxy-agent": "5.0.0",
     "inquirer": "8.0.0",
     "jsdom": "^16.5.3",
+    "koa": "^2.13.4",
+    "koa-bodyparser": "^4.3.0",
+    "koa-json": "^2.0.2",
     "lowdb": "1.0.0",
     "lunr": "^2.3.9",
-    "multer": "^1.4.3",
+    "multer": "^1.4.4",
```
[the remark this answers is not in the document]
Comment (kspearrin) [reply package.json:114]: I believe so, yes.

File: src/commands/serve.command.ts (hunk 1 above)
Comment (kspearrin) [reply src/commands/serve.command.ts:312]: Ideally koa would only activate the first, or most specific matching route, but that doesn't seem to be the case.

Comment (MGibson1) [reply src/commands/serve.command.ts:312]: What do you think about prepending the path with something to make the routes less generic? I'm thinking `/object/:object/...` would work for these examples.

I'm hesitant to committing to this path structure for a public API if we need this workaround

Comment (kspearrin) [reply src/commands/serve.command.ts:312]: That was one solution I thought about, but I was trying to present the routes in a very RESTful-like way. Using a "object" prefix could work. Let me think about that some more.

**Review by MGibson1 — DISMISSED** [review]

**Pull request merged**